improve qcat_nhwc performance on both multi-core and single-core

mingfeima opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #74999
* #68328
* #72575
* #71903
* __->__ #69667
* #69601
* #69600
* #69599
* #69598
* #69517

Differential Revision: [D33353155](https://our.internmc.facebook.com/intern/diff/D33353155)

mingfeima committed:
improve qcat_nhwc performance on both multi-core and single-core

[ghstack-poisoned]

mingfeima committed:
Update on "improve qcat_nhwc performance on both multi-core and single-core"




[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33353155).

mingfeima committed:
Update on "improve qcat_nhwc performance on both multi-core and single-core"


Differential Revision: [D33353155](https://our.internmc.facebook.com/intern/diff/D33353155)

[ghstack-poisoned]

mingfeima committed:
Update on "improve qcat_nhwc performance on both multi-core and single-core"


Differential Revision: [D33353155](https://our.internmc.facebook.com/intern/diff/D33353155)

[ghstack-poisoned]

mingfeima committed:
Update on "improve qcat_nhwc performance on both multi-core and single-core"


Differential Revision: [D33353155](https://our.internmc.facebook.com/intern/diff/D33353155)

[ghstack-poisoned]

Xia-Weiwen commented:
Hi @jerryzh168 We have a few int8-related PRs targeting 1.11. And this is one of them. Could you prioritize to review these PRs? Thanks.

mingfeima committed:
Update on "improve qcat_nhwc performance on both multi-core and single-core"


Differential Revision: [D33353155](https://our.internmc.facebook.com/intern/diff/D33353155)

[ghstack-poisoned]

mingfeima committed:
Update on "improve qcat_nhwc performance on both multi-core and single-core"


Differential Revision: [D33353155](https://our.internmc.facebook.com/intern/diff/D33353155)

[ghstack-poisoned]

mingfeima committed:
Update on "improve qcat_nhwc performance on both multi-core and single-core"


Differential Revision: [D33353155](https://our.internmc.facebook.com/intern/diff/D33353155)

[ghstack-poisoned]

mingfeima committed:
Update on "improve qcat_nhwc performance on both multi-core and single-core"


Differential Revision: [D33353155](https://our.internmc.facebook.com/intern/diff/D33353155)

[ghstack-poisoned]

mingfeima committed:
Update on "improve qcat_nhwc performance on both multi-core and single-core"


Differential Revision: [D33353155](https://our.internmc.facebook.com/intern/diff/D33353155)

[ghstack-poisoned]

mingfeima committed:
Update on "improve qcat_nhwc performance on both multi-core and single-core"


Differential Revision: [D33353155](https://our.internmc.facebook.com/intern/diff/D33353155)

[ghstack-poisoned]

mingfeima committed:
Update on "improve qcat_nhwc performance on both multi-core and single-core"


Differential Revision: [D33353155](https://our.internmc.facebook.com/intern/diff/D33353155)

[ghstack-poisoned]

dzdang commented:
Sorry for the delayed response. Could you please provide a high level summary of the changes made in this PR (as well as the other PRs) and I'll try to get this approved ASAP afterwards? Thanks.

mingfeima committed:
Update on "improve qcat_nhwc performance on both multi-core and single-core"


Differential Revision: [D33353155](https://our.internmc.facebook.com/intern/diff/D33353155)

[ghstack-poisoned]

mingfeima committed:
Update on "improve qcat_nhwc performance on both multi-core and single-core"


Differential Revision: [D33353155](https://our.internmc.facebook.com/intern/diff/D33353155)

[ghstack-poisoned]

mingfeima committed:
Update on "improve qcat_nhwc performance on both multi-core and single-core"


Differential Revision: [D33353155](https://our.internmc.facebook.com/intern/diff/D33353155)

[ghstack-poisoned]

mingfeima commented:
#### Updates of this PR

The original kernel on qcat has two branches:
* nhwc (dim=1): on physical memory, it equals to concat with input shape of {M, N} and dim = -1. (This is a fast path).
* else: dequant-cat-quant. (This is a slow path)

This PR updates in two ways for the nhwc fast path (the slow path is untouched)
* parallels on the outer dimension aka. nbatch * height * width
* vectorize remainder.

mingfeima commented:
#### 2.a) single core run on long tag
only 'contigall' + 'dim2' is affected, will go the fast path with #72575.


name | before | after | speedup
-- | -- | -- | --
qcat_M128_N128_K1_L5_dim0_contigall_dtypetorch.quint8 | 30.655 | 28.598 | 107.19%
qcat_M128_N128_K1_L5_dim0_contigone_dtypetorch.quint8 | 43.688 | 42.502 | 102.79%
qcat_M128_N128_K1_L5_dim0_contignone_dtypetorch.quint8 | 55.322 | 53.189 | 104.01%
qcat_M128_N128_K1_L5_dim1_contigall_dtypetorch.quint8 | 30.063 | 28.595 | 105.13%
qcat_M128_N128_K1_L5_dim1_contigone_dtypetorch.quint8 | 44.15 | 42.518 | 103.84%
qcat_M128_N128_K1_L5_dim1_contignone_dtypetorch.quint8 | 55.285 | 53.218 | 103.88%
qcat_M128_N128_K1_L5_dim2_contigall_dtypetorch.quint8 | 140.599 | 150.521 | 93.41%
qcat_M128_N128_K1_L5_dim2_contigone_dtypetorch.quint8 | 150.865 | 157.565 | 95.75%
qcat_M128_N128_K1_L5_dim2_contignone_dtypetorch.quint8 | 162.481 | 161.452 | 100.64%
qcat_M128_N128_K1_L7_dim0_contigall_dtypetorch.quint8 | 30.164 | 28.82 | 104.66%
qcat_M128_N128_K1_L7_dim0_contigone_dtypetorch.quint8 | 43.897 | 44.712 | 98.18%
qcat_M128_N128_K1_L7_dim0_contignone_dtypetorch.quint8 | 55.346 | 55.845 | 99.11%
qcat_M128_N128_K1_L7_dim1_contigall_dtypetorch.quint8 | 30.311 | 28.579 | 106.06%
qcat_M128_N128_K1_L7_dim1_contigone_dtypetorch.quint8 | 43.74 | 44.882 | 97.46%
qcat_M128_N128_K1_L7_dim1_contignone_dtypetorch.quint8 | 55.604 | 53.148 | 104.62%
qcat_M128_N128_K1_L7_dim2_contigall_dtypetorch.quint8 | 137.647 | 149.977 | 91.78%
qcat_M128_N128_K1_L7_dim2_contigone_dtypetorch.quint8 | 159.196 | 150.771 | 105.59%
qcat_M128_N128_K1_L7_dim2_contignone_dtypetorch.quint8 | 162.19 | 168.403 | 96.31%
qcat_M128_N128_K2_L5_dim0_contigall_dtypetorch.quint8 | 45.693 | 43.88 | 104.13%
qcat_M128_N128_K2_L5_dim0_contigone_dtypetorch.quint8 | 86.728 | 83.549 | 103.80%
qcat_M128_N128_K2_L5_dim0_contignone_dtypetorch.quint8 | 119.366 | 115.412 | 103.43%
qcat_M128_N128_K2_L5_dim1_contigall_dtypetorch.quint8 | 45.278 | 44.064 | 102.76%
qcat_M128_N128_K2_L5_dim1_contigone_dtypetorch.quint8 | 86.811 | 83.406 | 104.08%
qcat_M128_N128_K2_L5_dim1_contignone_dtypetorch.quint8 | 118.886 | 115.255 | 103.15%
qcat_M128_N128_K2_L5_dim2_contigall_dtypetorch.quint8 | 158.405 | 148.946 | 106.35%
qcat_M128_N128_K2_L5_dim2_contigone_dtypetorch.quint8 | 198.839 | 204.857 | 97.06%
qcat_M128_N128_K2_L5_dim2_contignone_dtypetorch.quint8 | 230.314 | 237.941 | 96.79%
qcat_M128_N128_K2_L7_dim0_contigall_dtypetorch.quint8 | 46.013 | 44.784 | 102.74%
qcat_M128_N128_K2_L7_dim0_contigone_dtypetorch.quint8 | 86.472 | 83.51 | 103.55%
qcat_M128_N128_K2_L7_dim0_contignone_dtypetorch.quint8 | 119.114 | 115.635 | 103.01%
qcat_M128_N128_K2_L7_dim1_contigall_dtypetorch.quint8 | 45.533 | 44.023 | 103.43%
qcat_M128_N128_K2_L7_dim1_contigone_dtypetorch.quint8 | 85.799 | 87.154 | 98.45%
qcat_M128_N128_K2_L7_dim1_contignone_dtypetorch.quint8 | 118.281 | 120.536 | 98.13%
qcat_M128_N128_K2_L7_dim2_contigall_dtypetorch.quint8 | 158.26 | 148.852 | 106.32%
qcat_M128_N128_K2_L7_dim2_contigone_dtypetorch.quint8 | 197.336 | 204.856 | 96.33%
qcat_M128_N128_K2_L7_dim2_contignone_dtypetorch.quint8 | 231.257 | 226.971 | 101.89%
qcat_M128_N1024_K1_L5_dim0_contigall_dtypetorch.quint8 | 216.022 | 213.553 | 101.16%
qcat_M128_N1024_K1_L5_dim0_contigone_dtypetorch.quint8 | 306.718 | 305.674 | 100.34%
qcat_M128_N1024_K1_L5_dim0_contignone_dtypetorch.quint8 | 380.211 | 380.617 | 99.89%
qcat_M128_N1024_K1_L5_dim1_contigall_dtypetorch.quint8 | 210.673 | 209.229 | 100.69%
qcat_M128_N1024_K1_L5_dim1_contigone_dtypetorch.quint8 | 302.956 | 309.099 | 98.01%
qcat_M128_N1024_K1_L5_dim1_contignone_dtypetorch.quint8 | 376.85 | 378.56 | 99.55%
qcat_M128_N1024_K1_L5_dim2_contigall_dtypetorch.quint8 | 1038.25 | 1097.58 | 94.59%
qcat_M128_N1024_K1_L5_dim2_contigone_dtypetorch.quint8 | 1130.294 | 1132.707 | 99.79%
qcat_M128_N1024_K1_L5_dim2_contignone_dtypetorch.quint8 | 1203.506 | 1207.006 | 99.71%
qcat_M128_N1024_K1_L7_dim0_contigall_dtypetorch.quint8 | 214.781 | 213.353 | 100.67%
qcat_M128_N1024_K1_L7_dim0_contigone_dtypetorch.quint8 | 306.589 | 305.422 | 100.38%
qcat_M128_N1024_K1_L7_dim0_contignone_dtypetorch.quint8 | 380.943 | 381.855 | 99.76%
qcat_M128_N1024_K1_L7_dim1_contigall_dtypetorch.quint8 | 210.618 | 209.323 | 100.62%
qcat_M128_N1024_K1_L7_dim1_contigone_dtypetorch.quint8 | 302.674 | 301.907 | 100.25%
qcat_M128_N1024_K1_L7_dim1_contignone_dtypetorch.quint8 | 376.76 | 379.328 | 99.32%
qcat_M128_N1024_K1_L7_dim2_contigall_dtypetorch.quint8 | 1038.254 | 1151.613 | 90.16%
qcat_M128_N1024_K1_L7_dim2_contigone_dtypetorch.quint8 | 1130.307 | 1131.907 | 99.86%
qcat_M128_N1024_K1_L7_dim2_contignone_dtypetorch.quint8 | 1202.381 | 1208.431 | 99.50%
qcat_M128_N1024_K2_L5_dim0_contigall_dtypetorch.quint8 | 402.73 | 399.064 | 100.92%
qcat_M128_N1024_K2_L5_dim0_contigone_dtypetorch.quint8 | 613.776 | 611.191 | 100.42%
qcat_M128_N1024_K2_L5_dim0_contignone_dtypetorch.quint8 | 819.832 | 817.502 | 100.29%
qcat_M128_N1024_K2_L5_dim1_contigall_dtypetorch.quint8 | 401.795 | 399.453 | 100.59%
qcat_M128_N1024_K2_L5_dim1_contigone_dtypetorch.quint8 | 614.097 | 611.619 | 100.41%
qcat_M128_N1024_K2_L5_dim1_contignone_dtypetorch.quint8 | 819.945 | 818.727 | 100.15%
qcat_M128_N1024_K2_L5_dim2_contigall_dtypetorch.quint8 | 1207.522 | 1118.871 | 107.92%
qcat_M128_N1024_K2_L5_dim2_contigone_dtypetorch.quint8 | 1419.29 | 1420.25 | 99.93%
qcat_M128_N1024_K2_L5_dim2_contignone_dtypetorch.quint8 | 1625.243 | 1626.659 | 99.91%
qcat_M128_N1024_K2_L7_dim0_contigall_dtypetorch.quint8 | 401.746 | 400.014 | 100.43%
qcat_M128_N1024_K2_L7_dim0_contigone_dtypetorch.quint8 | 614.231 | 611.466 | 100.45%
qcat_M128_N1024_K2_L7_dim0_contignone_dtypetorch.quint8 | 824.294 | 818.049 | 100.76%
qcat_M128_N1024_K2_L7_dim1_contigall_dtypetorch.quint8 | 403.055 | 400.02 | 100.76%
qcat_M128_N1024_K2_L7_dim1_contigone_dtypetorch.quint8 | 613.994 | 611.549 | 100.40%
qcat_M128_N1024_K2_L7_dim1_contignone_dtypetorch.quint8 | 820.629 | 819.402 | 100.15%
qcat_M128_N1024_K2_L7_dim2_contigall_dtypetorch.quint8 | 1206.372 | 1119.401 | 107.77%
qcat_M128_N1024_K2_L7_dim2_contigone_dtypetorch.quint8 | 1419.555 | 1422.532 | 99.79%
qcat_M128_N1024_K2_L7_dim2_contignone_dtypetorch.quint8 | 1624.712 | 1628.434 | 99.77%
qcat_M1024_N128_K1_L5_dim0_contigall_dtypetorch.quint8 | 214.901 | 213.206 | 100.80%
qcat_M1024_N128_K1_L5_dim0_contigone_dtypetorch.quint8 | 320.47 | 319.518 | 100.30%
qcat_M1024_N128_K1_L5_dim0_contignone_dtypetorch.quint8 | 405.489 | 410.412 | 98.80%
qcat_M1024_N128_K1_L5_dim1_contigall_dtypetorch.quint8 | 210.631 | 209.601 | 100.49%
qcat_M1024_N128_K1_L5_dim1_contigone_dtypetorch.quint8 | 317.207 | 316.037 | 100.37%
qcat_M1024_N128_K1_L5_dim1_contignone_dtypetorch.quint8 | 403.187 | 406.763 | 99.12%
qcat_M1024_N128_K1_L5_dim2_contigall_dtypetorch.quint8 | 1039.052 | 1097.421 | 94.68%
qcat_M1024_N128_K1_L5_dim2_contigone_dtypetorch.quint8 | 1143.053 | 1199.613 | 95.29%
qcat_M1024_N128_K1_L5_dim2_contignone_dtypetorch.quint8 | 1228.444 | 1293.647 | 94.96%
qcat_M1024_N128_K1_L7_dim0_contigall_dtypetorch.quint8 | 215.336 | 213.366 | 100.92%
qcat_M1024_N128_K1_L7_dim0_contigone_dtypetorch.quint8 | 322.358 | 319.545 | 100.88%
qcat_M1024_N128_K1_L7_dim0_contignone_dtypetorch.quint8 | 407.653 | 410.32 | 99.35%
qcat_M1024_N128_K1_L7_dim1_contigall_dtypetorch.quint8 | 210.893 | 209.503 | 100.66%
qcat_M1024_N128_K1_L7_dim1_contigone_dtypetorch.quint8 | 317.695 | 324.64 | 97.86%
qcat_M1024_N128_K1_L7_dim1_contignone_dtypetorch.quint8 | 402.159 | 407.166 | 98.77%
qcat_M1024_N128_K1_L7_dim2_contigall_dtypetorch.quint8 | 1038.747 | 1097.86 | 94.62%
qcat_M1024_N128_K1_L7_dim2_contigone_dtypetorch.quint8 | 1144.022 | 1147.607 | 99.69%
qcat_M1024_N128_K1_L7_dim2_contignone_dtypetorch.quint8 | 1227.044 | 1236.845 | 99.21%
qcat_M1024_N128_K2_L5_dim0_contigall_dtypetorch.quint8 | 401.612 | 399.323 | 100.57%
qcat_M1024_N128_K2_L5_dim0_contigone_dtypetorch.quint8 | 645.033 | 639.345 | 100.89%
qcat_M1024_N128_K2_L5_dim0_contignone_dtypetorch.quint8 | 872.447 | 866.85 | 100.65%
qcat_M1024_N128_K2_L5_dim1_contigall_dtypetorch.quint8 | 402.559 | 400.078 | 100.62%
qcat_M1024_N128_K2_L5_dim1_contigone_dtypetorch.quint8 | 645.366 | 637.191 | 101.28%
qcat_M1024_N128_K2_L5_dim1_contignone_dtypetorch.quint8 | 873.022 | 866.35 | 100.77%
qcat_M1024_N128_K2_L5_dim2_contigall_dtypetorch.quint8 | 1206.768 | 1118.125 | 107.93%
qcat_M1024_N128_K2_L5_dim2_contigone_dtypetorch.quint8 | 1450.632 | 1484.15 | 97.74%
qcat_M1024_N128_K2_L5_dim2_contignone_dtypetorch.quint8 | 1684.167 | 1680.631 | 100.21%
qcat_M1024_N128_K2_L7_dim0_contigall_dtypetorch.quint8 | 401.589 | 399.442 | 100.54%
qcat_M1024_N128_K2_L7_dim0_contigone_dtypetorch.quint8 | 644.979 | 636.695 | 101.30%
qcat_M1024_N128_K2_L7_dim0_contignone_dtypetorch.quint8 | 872.556 | 866.919 | 100.65%
qcat_M1024_N128_K2_L7_dim1_contigall_dtypetorch.quint8 | 401.641 | 399.843 | 100.45%
qcat_M1024_N128_K2_L7_dim1_contigone_dtypetorch.quint8 | 644.96 | 638.121 | 101.07%
qcat_M1024_N128_K2_L7_dim1_contignone_dtypetorch.quint8 | 872.391 | 868.398 | 100.46%
qcat_M1024_N128_K2_L7_dim2_contigall_dtypetorch.quint8 | 1206.927 | 1119.085 | 107.85%
qcat_M1024_N128_K2_L7_dim2_contigone_dtypetorch.quint8 | 1450.568 | 1445.346 | 100.36%
qcat_M1024_N128_K2_L7_dim2_contignone_dtypetorch.quint8 | 1683.981 | 1683.122 | 100.05%
qcat_M1024_N1024_K1_L5_dim0_contigall_dtypetorch.quint8 | 1554.196 | 1563.52 | 99.40%
qcat_M1024_N1024_K1_L5_dim0_contigone_dtypetorch.quint8 | 2355.793 | 2458.883 | 95.81%
qcat_M1024_N1024_K1_L5_dim0_contignone_dtypetorch.quint8 | 3163.628 | 3225.432 | 98.08%
qcat_M1024_N1024_K1_L5_dim1_contigall_dtypetorch.quint8 | 1562.866 | 1559.673 | 100.20%
qcat_M1024_N1024_K1_L5_dim1_contigone_dtypetorch.quint8 | 2353.879 | 2455.096 | 95.88%
qcat_M1024_N1024_K1_L5_dim1_contignone_dtypetorch.quint8 | 3161.755 | 3249.027 | 97.31%
qcat_M1024_N1024_K1_L5_dim2_contigall_dtypetorch.quint8 | 8090.493 | 8708.694 | 92.90%
qcat_M1024_N1024_K1_L5_dim2_contigone_dtypetorch.quint8 | 8886.6 | 8970.501 | 99.06%
qcat_M1024_N1024_K1_L5_dim2_contignone_dtypetorch.quint8 | 9703.813 | 9773.119 | 99.29%
qcat_M1024_N1024_K1_L7_dim0_contigall_dtypetorch.quint8 | 1558.128 | 1559.422 | 99.92%
qcat_M1024_N1024_K1_L7_dim0_contigone_dtypetorch.quint8 | 2354.07 | 2451.17 | 96.04%
qcat_M1024_N1024_K1_L7_dim0_contignone_dtypetorch.quint8 | 3161.92 | 3253.686 | 97.18%
qcat_M1024_N1024_K1_L7_dim1_contigall_dtypetorch.quint8 | 1562.816 | 1558.354 | 100.29%
qcat_M1024_N1024_K1_L7_dim1_contigone_dtypetorch.quint8 | 2353.119 | 2455.177 | 95.84%
qcat_M1024_N1024_K1_L7_dim1_contignone_dtypetorch.quint8 | 3159.771 | 3246.586 | 97.33%
qcat_M1024_N1024_K1_L7_dim2_contigall_dtypetorch.quint8 | 8095.204 | 8718.785 | 92.85%
qcat_M1024_N1024_K1_L7_dim2_contigone_dtypetorch.quint8 | 8880.988 | 8982.98 | 98.86%
qcat_M1024_N1024_K1_L7_dim2_contignone_dtypetorch.quint8 | 9692.385 | 9773.33 | 99.17%
qcat_M1024_N1024_K2_L5_dim0_contigall_dtypetorch.quint8 | 4526.41 | 4621.931 | 97.93%
qcat_M1024_N1024_K2_L5_dim0_contigone_dtypetorch.quint8 | 11622.44 | 11770.23 | 98.74%
qcat_M1024_N1024_K2_L5_dim0_contignone_dtypetorch.quint8 | 18363.29 | 18372.8 | 99.95%
qcat_M1024_N1024_K2_L5_dim1_contigall_dtypetorch.quint8 | 4496.932 | 4561.083 | 98.59%
qcat_M1024_N1024_K2_L5_dim1_contigone_dtypetorch.quint8 | 11604.65 | 11689.68 | 99.27%
qcat_M1024_N1024_K2_L5_dim1_contignone_dtypetorch.quint8 | 18226.98 | 18308.03 | 99.56%
qcat_M1024_N1024_K2_L5_dim2_contigall_dtypetorch.quint8 | 10523.49 | 8832.709 | 119.14%
qcat_M1024_N1024_K2_L5_dim2_contigone_dtypetorch.quint8 | 17498.36 | 17568.24 | 99.60%
qcat_M1024_N1024_K2_L5_dim2_contignone_dtypetorch.quint8 | 24065.66 | 24124.24 | 99.76%
qcat_M1024_N1024_K2_L7_dim0_contigall_dtypetorch.quint8 | 4537.342 | 4622.38 | 98.16%
qcat_M1024_N1024_K2_L7_dim0_contigone_dtypetorch.quint8 | 11632.24 | 11714.44 | 99.30%
qcat_M1024_N1024_K2_L7_dim0_contignone_dtypetorch.quint8 | 18306.55 | 18367.54 | 99.67%
qcat_M1024_N1024_K2_L7_dim1_contigall_dtypetorch.quint8 | 4501.414 | 4524.194 | 99.50%
qcat_M1024_N1024_K2_L7_dim1_contigone_dtypetorch.quint8 | 11625.88 | 11652.65 | 99.77%
qcat_M1024_N1024_K2_L7_dim1_contignone_dtypetorch.quint8 | 18269.28 | 18277.81 | 99.95%
qcat_M1024_N1024_K2_L7_dim2_contigall_dtypetorch.quint8 | 10505.9 | 8829.414 | 118.99%
qcat_M1024_N1024_K2_L7_dim2_contigone_dtypetorch.quint8 | 17507.19 | 17533.27 | 99.85%
qcat_M1024_N1024_K2_L7_dim2_contignone_dtypetorch.quint8 | 24055.61 | 24117.45 | 99.74%

mingfeima commented:
#### 2.b) single socket run on long tag
only 'contigall' + 'dim2' is affected, will go the fast path with #72575.

name | before | after | speedup
-- | -- | -- | --
qcat_M128_N128_K1_L5_dim0_contigall_dtypetorch.quint8 | 66.625 | 65.314 | 102.01%
qcat_M128_N128_K1_L5_dim0_contigone_dtypetorch.quint8 | 91.982 | 89.569 | 102.69%
qcat_M128_N128_K1_L5_dim0_contignone_dtypetorch.quint8 | 111.136 | 108.351 | 102.57%
qcat_M128_N128_K1_L5_dim1_contigall_dtypetorch.quint8 | 68.137 | 67.273 | 101.28%
qcat_M128_N128_K1_L5_dim1_contigone_dtypetorch.quint8 | 92.802 | 91.655 | 101.25%
qcat_M128_N128_K1_L5_dim1_contignone_dtypetorch.quint8 | 113.045 | 110.073 | 102.70%
qcat_M128_N128_K1_L5_dim2_contigall_dtypetorch.quint8 | 74.955 | 25.736 | 291.25%
qcat_M128_N128_K1_L5_dim2_contigone_dtypetorch.quint8 | 101.499 | 100.443 | 101.05%
qcat_M128_N128_K1_L5_dim2_contignone_dtypetorch.quint8 | 119.178 | 116.479 | 102.32%
qcat_M128_N128_K1_L7_dim0_contigall_dtypetorch.quint8 | 66.382 | 65.996 | 100.58%
qcat_M128_N128_K1_L7_dim0_contigone_dtypetorch.quint8 | 92.916 | 90.189 | 103.02%
qcat_M128_N128_K1_L7_dim0_contignone_dtypetorch.quint8 | 111.123 | 107.743 | 103.14%
qcat_M128_N128_K1_L7_dim1_contigall_dtypetorch.quint8 | 68.726 | 67.543 | 101.75%
qcat_M128_N128_K1_L7_dim1_contigone_dtypetorch.quint8 | 94.933 | 90.47 | 104.93%
qcat_M128_N128_K1_L7_dim1_contignone_dtypetorch.quint8 | 113.101 | 109.056 | 103.71%
qcat_M128_N128_K1_L7_dim2_contigall_dtypetorch.quint8 | 74.834 | 25.429 | 294.29%
qcat_M128_N128_K1_L7_dim2_contigone_dtypetorch.quint8 | 99.303 | 98.273 | 101.05%
qcat_M128_N128_K1_L7_dim2_contignone_dtypetorch.quint8 | 118.909 | 115.675 | 102.80%
qcat_M128_N128_K2_L5_dim0_contigall_dtypetorch.quint8 | 98.171 | 97.15 | 101.05%
qcat_M128_N128_K2_L5_dim0_contigone_dtypetorch.quint8 | 162.491 | 159.682 | 101.76%
qcat_M128_N128_K2_L5_dim0_contignone_dtypetorch.quint8 | 213.87 | 208.543 | 102.55%
qcat_M128_N128_K2_L5_dim1_contigall_dtypetorch.quint8 | 102.091 | 100.343 | 101.74%
qcat_M128_N128_K2_L5_dim1_contigone_dtypetorch.quint8 | 165.392 | 163.035 | 101.45%
qcat_M128_N128_K2_L5_dim1_contignone_dtypetorch.quint8 | 216.529 | 211.66 | 102.30%
qcat_M128_N128_K2_L5_dim2_contigall_dtypetorch.quint8 | 189.98 | 26.041 | 729.54%
qcat_M128_N128_K2_L5_dim2_contigone_dtypetorch.quint8 | 252.409 | 261.298 | 96.60%
qcat_M128_N128_K2_L5_dim2_contignone_dtypetorch.quint8 | 302.33 | 306.116 | 98.76%
qcat_M128_N128_K2_L7_dim0_contigall_dtypetorch.quint8 | 98.038 | 96.739 | 101.34%
qcat_M128_N128_K2_L7_dim0_contigone_dtypetorch.quint8 | 162.527 | 160.053 | 101.55%
qcat_M128_N128_K2_L7_dim0_contignone_dtypetorch.quint8 | 213.324 | 208.526 | 102.30%
qcat_M128_N128_K2_L7_dim1_contigall_dtypetorch.quint8 | 101.609 | 100.187 | 101.42%
qcat_M128_N128_K2_L7_dim1_contigone_dtypetorch.quint8 | 165.205 | 163.147 | 101.26%
qcat_M128_N128_K2_L7_dim1_contignone_dtypetorch.quint8 | 217.101 | 211.62 | 102.59%
qcat_M128_N128_K2_L7_dim2_contigall_dtypetorch.quint8 | 189.387 | 25.832 | 733.15%
qcat_M128_N128_K2_L7_dim2_contigone_dtypetorch.quint8 | 252.561 | 262.409 | 96.25%
qcat_M128_N128_K2_L7_dim2_contignone_dtypetorch.quint8 | 301.584 | 306.3 | 98.46%
qcat_M128_N1024_K1_L5_dim0_contigall_dtypetorch.quint8 | 101.957 | 105.653 | 96.50%
qcat_M128_N1024_K1_L5_dim0_contigone_dtypetorch.quint8 | 155.669 | 157.421 | 98.89%
qcat_M128_N1024_K1_L5_dim0_contignone_dtypetorch.quint8 | 202.582 | 203.506 | 99.55%
qcat_M128_N1024_K1_L5_dim1_contigall_dtypetorch.quint8 | 103.992 | 105.183 | 98.87%
qcat_M128_N1024_K1_L5_dim1_contigone_dtypetorch.quint8 | 157.384 | 157.337 | 100.03%
qcat_M128_N1024_K1_L5_dim1_contignone_dtypetorch.quint8 | 203.978 | 204.283 | 99.85%
qcat_M128_N1024_K1_L5_dim2_contigall_dtypetorch.quint8 | 116.568 | 86.399 | 134.92%
qcat_M128_N1024_K1_L5_dim2_contigone_dtypetorch.quint8 | 172.282 | 173.001 | 99.58%
qcat_M128_N1024_K1_L5_dim2_contignone_dtypetorch.quint8 | 220.45 | 220.848 | 99.82%
qcat_M128_N1024_K1_L7_dim0_contigall_dtypetorch.quint8 | 102.046 | 104.923 | 97.26%
qcat_M128_N1024_K1_L7_dim0_contigone_dtypetorch.quint8 | 155.307 | 157.823 | 98.41%
qcat_M128_N1024_K1_L7_dim0_contignone_dtypetorch.quint8 | 201.965 | 202.872 | 99.55%
qcat_M128_N1024_K1_L7_dim1_contigall_dtypetorch.quint8 | 104.15 | 105.166 | 99.03%
qcat_M128_N1024_K1_L7_dim1_contigone_dtypetorch.quint8 | 156.693 | 157.924 | 99.22%
qcat_M128_N1024_K1_L7_dim1_contignone_dtypetorch.quint8 | 203.981 | 203.64 | 100.17%
qcat_M128_N1024_K1_L7_dim2_contigall_dtypetorch.quint8 | 116.174 | 86.617 | 134.12%
qcat_M128_N1024_K1_L7_dim2_contigone_dtypetorch.quint8 | 172.218 | 172.927 | 99.59%
qcat_M128_N1024_K1_L7_dim2_contignone_dtypetorch.quint8 | 220.705 | 220.146 | 100.25%
qcat_M128_N1024_K2_L5_dim0_contigall_dtypetorch.quint8 | 110.313 | 108.401 | 101.76%
qcat_M128_N1024_K2_L5_dim0_contigone_dtypetorch.quint8 | 173.462 | 170.985 | 101.45%
qcat_M128_N1024_K2_L5_dim0_contignone_dtypetorch.quint8 | 229.565 | 227.859 | 100.75%
qcat_M128_N1024_K2_L5_dim1_contigall_dtypetorch.quint8 | 110.032 | 109.765 | 100.24%
qcat_M128_N1024_K2_L5_dim1_contigone_dtypetorch.quint8 | 172.898 | 170.692 | 101.29%
qcat_M128_N1024_K2_L5_dim1_contignone_dtypetorch.quint8 | 229.355 | 228.022 | 100.58%
qcat_M128_N1024_K2_L5_dim2_contigall_dtypetorch.quint8 | 206.839 | 87.872 | 235.39%
qcat_M128_N1024_K2_L5_dim2_contigone_dtypetorch.quint8 | 273.416 | 281.554 | 97.11%
qcat_M128_N1024_K2_L5_dim2_contignone_dtypetorch.quint8 | 323.138 | 340.574 | 94.88%
qcat_M128_N1024_K2_L7_dim0_contigall_dtypetorch.quint8 | 110.33 | 107.791 | 102.36%
qcat_M128_N1024_K2_L7_dim0_contigone_dtypetorch.quint8 | 172.948 | 170.49 | 101.44%
qcat_M128_N1024_K2_L7_dim0_contignone_dtypetorch.quint8 | 229.566 | 227.204 | 101.04%
qcat_M128_N1024_K2_L7_dim1_contigall_dtypetorch.quint8 | 110.912 | 110.515 | 100.36%
qcat_M128_N1024_K2_L7_dim1_contigone_dtypetorch.quint8 | 173.059 | 171.273 | 101.04%
qcat_M128_N1024_K2_L7_dim1_contignone_dtypetorch.quint8 | 230.456 | 228.543 | 100.84%
qcat_M128_N1024_K2_L7_dim2_contigall_dtypetorch.quint8 | 206.742 | 87.549 | 236.14%
qcat_M128_N1024_K2_L7_dim2_contigone_dtypetorch.quint8 | 273.089 | 281.643 | 96.96%
qcat_M128_N1024_K2_L7_dim2_contignone_dtypetorch.quint8 | 323.406 | 341.095 | 94.81%
qcat_M1024_N128_K1_L5_dim0_contigall_dtypetorch.quint8 | 101.858 | 104.309 | 97.65%
qcat_M1024_N128_K1_L5_dim0_contigone_dtypetorch.quint8 | 158.912 | 161.156 | 98.61%
qcat_M1024_N128_K1_L5_dim0_contignone_dtypetorch.quint8 | 211.024 | 211.87 | 99.60%
qcat_M1024_N128_K1_L5_dim1_contigall_dtypetorch.quint8 | 106.078 | 106.431 | 99.67%
qcat_M1024_N128_K1_L5_dim1_contigone_dtypetorch.quint8 | 162.717 | 163.404 | 99.58%
qcat_M1024_N128_K1_L5_dim1_contignone_dtypetorch.quint8 | 213.873 | 214.139 | 99.88%
qcat_M1024_N128_K1_L5_dim2_contigall_dtypetorch.quint8 | 116.152 | 86.787 | 133.84%
qcat_M1024_N128_K1_L5_dim2_contigone_dtypetorch.quint8 | 175.229 | 176.397 | 99.34%
qcat_M1024_N128_K1_L5_dim2_contignone_dtypetorch.quint8 | 228.105 | 230.731 | 98.86%
qcat_M1024_N128_K1_L7_dim0_contigall_dtypetorch.quint8 | 101.748 | 104.829 | 97.06%
qcat_M1024_N128_K1_L7_dim0_contigone_dtypetorch.quint8 | 158.436 | 161.161 | 98.31%
qcat_M1024_N128_K1_L7_dim0_contignone_dtypetorch.quint8 | 210.126 | 210.907 | 99.63%
qcat_M1024_N128_K1_L7_dim1_contigall_dtypetorch.quint8 | 105.957 | 106.639 | 99.36%
qcat_M1024_N128_K1_L7_dim1_contigone_dtypetorch.quint8 | 162.615 | 163.918 | 99.21%
qcat_M1024_N128_K1_L7_dim1_contignone_dtypetorch.quint8 | 214.646 | 214.93 | 99.87%
qcat_M1024_N128_K1_L7_dim2_contigall_dtypetorch.quint8 | 116.217 | 86.401 | 134.51%
qcat_M1024_N128_K1_L7_dim2_contigone_dtypetorch.quint8 | 174.84 | 176.096 | 99.29%
qcat_M1024_N128_K1_L7_dim2_contignone_dtypetorch.quint8 | 228.7 | 229.916 | 99.47%
qcat_M1024_N128_K2_L5_dim0_contigall_dtypetorch.quint8 | 110.157 | 108.012 | 101.99%
qcat_M1024_N128_K2_L5_dim0_contigone_dtypetorch.quint8 | 174.995 | 171.051 | 102.31%
qcat_M1024_N128_K2_L5_dim0_contignone_dtypetorch.quint8 | 238.396 | 233.461 | 102.11%
qcat_M1024_N128_K2_L5_dim1_contigall_dtypetorch.quint8 | 112.44 | 113.892 | 98.73%
qcat_M1024_N128_K2_L5_dim1_contigone_dtypetorch.quint8 | 176.247 | 172.476 | 102.19%
qcat_M1024_N128_K2_L5_dim1_contignone_dtypetorch.quint8 | 239.644 | 234.955 | 102.00%
qcat_M1024_N128_K2_L5_dim2_contigall_dtypetorch.quint8 | 206.332 | 87.386 | 236.12%
qcat_M1024_N128_K2_L5_dim2_contigone_dtypetorch.quint8 | 275.015 | 282.15 | 97.47%
qcat_M1024_N128_K2_L5_dim2_contignone_dtypetorch.quint8 | 333.802 | 346.918 | 96.22%
qcat_M1024_N128_K2_L7_dim0_contigall_dtypetorch.quint8 | 110.125 | 107.784 | 102.17%
qcat_M1024_N128_K2_L7_dim0_contigone_dtypetorch.quint8 | 174.895 | 171.137 | 102.20%
qcat_M1024_N128_K2_L7_dim0_contignone_dtypetorch.quint8 | 238.038 | 233.046 | 102.14%
qcat_M1024_N128_K2_L7_dim1_contigall_dtypetorch.quint8 | 112.674 | 111.752 | 100.83%
qcat_M1024_N128_K2_L7_dim1_contigone_dtypetorch.quint8 | 176.349 | 173.093 | 101.88%
qcat_M1024_N128_K2_L7_dim1_contignone_dtypetorch.quint8 | 239.781 | 235.448 | 101.84%
qcat_M1024_N128_K2_L7_dim2_contigall_dtypetorch.quint8 | 206.509 | 88.256 | 233.99%
qcat_M1024_N128_K2_L7_dim2_contigone_dtypetorch.quint8 | 275.087 | 281.585 | 97.69%
qcat_M1024_N128_K2_L7_dim2_contignone_dtypetorch.quint8 | 334.996 | 346.379 | 96.71%
qcat_M1024_N1024_K1_L5_dim0_contigall_dtypetorch.quint8 | 180.802 | 180.517 | 100.16%
qcat_M1024_N1024_K1_L5_dim0_contigone_dtypetorch.quint8 | 250.042 | 251.739 | 99.33%
qcat_M1024_N1024_K1_L5_dim0_contignone_dtypetorch.quint8 | 314.521 | 313.854 | 100.21%
qcat_M1024_N1024_K1_L5_dim1_contigall_dtypetorch.quint8 | 128.148 | 137.424 | 93.25%
qcat_M1024_N1024_K1_L5_dim1_contigone_dtypetorch.quint8 | 204.156 | 204.754 | 99.71%
qcat_M1024_N1024_K1_L5_dim1_contignone_dtypetorch.quint8 | 264.958 | 265.494 | 99.80%
qcat_M1024_N1024_K1_L5_dim2_contigall_dtypetorch.quint8 | 177.399 | 589.022 | 30.12%
qcat_M1024_N1024_K1_L5_dim2_contigone_dtypetorch.quint8 | 252.521 | 269.215 | 93.80%
qcat_M1024_N1024_K1_L5_dim2_contignone_dtypetorch.quint8 | 315.085 | 314.672 | 100.13%
qcat_M1024_N1024_K1_L7_dim0_contigall_dtypetorch.quint8 | 179.251 | 178.888 | 100.20%
qcat_M1024_N1024_K1_L7_dim0_contigone_dtypetorch.quint8 | 250.48 | 250.31 | 100.07%
qcat_M1024_N1024_K1_L7_dim0_contignone_dtypetorch.quint8 | 315.102 | 313.474 | 100.52%
qcat_M1024_N1024_K1_L7_dim1_contigall_dtypetorch.quint8 | 127.943 | 127.854 | 100.07%
qcat_M1024_N1024_K1_L7_dim1_contigone_dtypetorch.quint8 | 204.089 | 204.544 | 99.78%
qcat_M1024_N1024_K1_L7_dim1_contignone_dtypetorch.quint8 | 264.869 | 266.092 | 99.54%
qcat_M1024_N1024_K1_L7_dim2_contigall_dtypetorch.quint8 | 176.178 | 590.36 | 29.84%
qcat_M1024_N1024_K1_L7_dim2_contigone_dtypetorch.quint8 | 252.455 | 269.445 | 93.69%
qcat_M1024_N1024_K1_L7_dim2_contignone_dtypetorch.quint8 | 315.412 | 315.505 | 99.97%
qcat_M1024_N1024_K2_L5_dim0_contigall_dtypetorch.quint8 | 382.51 | 377.698 | 101.27%
qcat_M1024_N1024_K2_L5_dim0_contigone_dtypetorch.quint8 | 765.963 | 761.428 | 100.60%
qcat_M1024_N1024_K2_L5_dim0_contignone_dtypetorch.quint8 | 1071.728 | 1078.052 | 99.41%
qcat_M1024_N1024_K2_L5_dim1_contigall_dtypetorch.quint8 | 341.568 | 349.344 | 97.77%
qcat_M1024_N1024_K2_L5_dim1_contigone_dtypetorch.quint8 | 712.698 | 722.768 | 98.61%
qcat_M1024_N1024_K2_L5_dim1_contignone_dtypetorch.quint8 | 985.522 | 1012.074 | 97.38%
qcat_M1024_N1024_K2_L5_dim2_contigall_dtypetorch.quint8 | 613.37 | 614.149 | 99.87%
qcat_M1024_N1024_K2_L5_dim2_contigone_dtypetorch.quint8 | 988.709 | 1021.438 | 96.80%
qcat_M1024_N1024_K2_L5_dim2_contignone_dtypetorch.quint8 | 1315.397 | 1369.855 | 96.02%
qcat_M1024_N1024_K2_L7_dim0_contigall_dtypetorch.quint8 | 385.041 | 385.925 | 99.77%
qcat_M1024_N1024_K2_L7_dim0_contigone_dtypetorch.quint8 | 762.838 | 760.816 | 100.27%
qcat_M1024_N1024_K2_L7_dim0_contignone_dtypetorch.quint8 | 1070.91 | 1079.429 | 99.21%
qcat_M1024_N1024_K2_L7_dim1_contigall_dtypetorch.quint8 | 343.694 | 351.932 | 97.66%
qcat_M1024_N1024_K2_L7_dim1_contigone_dtypetorch.quint8 | 714.776 | 725.277 | 98.55%
qcat_M1024_N1024_K2_L7_dim1_contignone_dtypetorch.quint8 | 988.816 | 1014.636 | 97.46%
qcat_M1024_N1024_K2_L7_dim2_contigall_dtypetorch.quint8 | 614.173 | 616.173 | 99.68%
qcat_M1024_N1024_K2_L7_dim2_contigone_dtypetorch.quint8 | 990.539 | 1024.585 | 96.68%
qcat_M1024_N1024_K2_L7_dim2_contignone_dtypetorch.quint8 | 1318.069 | 1375.259 | 95.84%

mingfeima commented:
#### 3. end to end model

pick googlenet and inceptionv3 from https://pytorch.org/vision/stable/models.html#quantized-models

* single core run
```bash
### before
### using OMP_NUM_THREADS=1
### using numactl --physcpubind=0-0 --membind=0

Running on device:  Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz

googlenet           : int8      : latency:      13.74 ms
inception_v3        : int8      : latency:      28.43 ms

### after
### using OMP_NUM_THREADS=1
### using numactl --physcpubind=0-0 --membind=0

Running on device:  Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz

googlenet           : int8      : latency:      12.12 ms
inception_v3        : int8      : latency:      23.29 ms
```

* 4 core run
```bash
### before
### using OMP_NUM_THREADS=4
### using numactl --physcpubind=0-3 --membind=0

Running on device:  Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz

googlenet           : int8      : latency:       7.73 ms
inception_v3        : int8      : latency:      14.86 ms

### after
### using OMP_NUM_THREADS=4
### using numactl --physcpubind=0-3 --membind=0

Running on device:  Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz

googlenet           : int8      : latency:       6.89 ms
inception_v3        : int8      : latency:      12.29 ms
```

mingfeima committed:
Update on "improve qcat_nhwc performance on both multi-core and single-core"


Differential Revision: [D33353155](https://our.internmc.facebook.com/intern/diff/D33353155)

[ghstack-poisoned]

mingfeima committed:
Update on "improve qcat_nhwc performance on both multi-core and single-core"


Differential Revision: [D33353155](https://our.internmc.facebook.com/intern/diff/D33353155)

[ghstack-poisoned]

dzdang reviewed (approved):
@mingfeima 
I approved all of these PRs and I'll try getting them landed internally ASAP

mingfeima committed:
Update on "improve qcat_nhwc performance on both multi-core and single-core"


Differential Revision: [D33353155](https://our.internmc.facebook.com/intern/diff/D33353155)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33353155).

mingfeima committed:
Update on "improve qcat_nhwc performance on both multi-core and single-core"


Differential Revision: [D33353155](https://our.internmc.facebook.com/intern/diff/D33353155)

[ghstack-poisoned]

mingfeima committed:
Update on "improve qcat_nhwc performance on both multi-core and single-core"


Differential Revision: [D33353155](https://our.internmc.facebook.com/intern/diff/D33353155)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33353155).
